Mode overhaul

Ambrevar opened the pull request:
Fixes #1836, #1005, #1198 and #2133.

To do:
- [x] Test auto-mode.
- [x] Address #1875.
- [x] Implement the mode togglers as in #1198.  But command creation is only done with macros, there is no functional alternative at the moment.  Can we create one?  Maybe leverage `fdefinition`.
  - [x] ~~`define-command` and co. should not call `defun`, `defparameter` if the target class has `toggler-command-p` set to nil.~~  Makes no sense, I was tired when writing this.
- [x] Disable mode togglers for the modes that don't need it, e.g. download-mode.
- [x] Fix the `sexp` slot in `command`.  See #2133.
- [x] Replace `format-mode` with a proper handling of `*print-escape*` in `print-object`.
- [ ] Subclass `command` for panels, bookmarklets, etc. to reduce macrology in the commands listed in https://github.com/atlas-engineer/nyxt/pull/2200#issuecomment-1105449923.
- [x] Add function / command to toggle mode, leverage `activate`.  Should we reintroduce the _destruction_ of modes?  For now we can only destroy a mode by destroying its buffer.  Note that to destroy a buffer, we should remove it from the "recent buffer" list.

Questions:

- [x] `enable-modes` and `disable-modes` have names similar to the commands `enable-mode` and `disable-mode`, but they are semantically very different.  Name suggestion?
  For now, I've opted to merge them.
- [x] Should we really use "mode shortcuts" (e.g. `emacs-mode` instead of `nyxt/emacs-mode:emacs-mode`)  I'd say no, just allow keywords to find the right mode in `current-mode`, it's useful from the REPL.
  I've also added @aartaka's `resolve-symbol`.
- [x] Shall we remove `rememberable-p` now that we have the metaclass option `toggler-command-p`?  I'd say yes, it's a trivial change in `auto-mode`.  Then we could renambe `toggler-command-p` to `togglable-p`.
  Artyom said no, as they are semantically different.

@jmercouris @aartaka Thoughts?

After merging, open issues about the questions listed in https://github.com/atlas-engineer/nyxt/pull/2200#issuecomment-1110800424.

Ambrevar commented:
I've just pushed a major overhaul of our command definitions.

Previously, it was hard to define commands in a functional manner, we strongly depended on the `define-command`, `make-command`, etc. macros.
This made it difficult to generate the mode togglers on demand.

Furthermore, there was a discrepancy between the side effects of `make-instance 'command`, `make-command` and `define-command`.  In particular, `make-command` used to _not_ trigger the hooks!

Should be all fixed now with massively reduced macrology.
The API is still the same with a minor difference, the `global-p` slot has been replaced by `visibility`.

To do:

- Fix the missing `sexp` slot.  Should we remove it?  Probably, now that we are shipping the source code.

Ambrevar commented:
Another thought: further macrology could be removed from
-  `define-internal-page-command-global`
-  `define-internal-page-command` 
-  `define-input-edit-command`
-  `define-bookmarklet-command`
-  `define-command-with-selection`
- `define-panel`
- `define-panel-global`

by subclassing `command`.  As usual, leveraging CLOS often proves a better idea than writing our own DSL.  Thoughts? @aartaka

Ambrevar commented:
Question: What about these "mode symbol shortcuts" that we have?  On `master` we allow `'web-mode` to stand for `nyxt/web-mode:web-mode` for instance.  

This makes for confusing code at times.  What if we didn't allow this and only passed the full symbol at all times?

The only place where the shortcut would make sense is in `current-mode`, but then I suggest to use a keyword to avoid interning an unbound symbol in the packages.  

Example: `(current-mode :web)` would return the appropriate mode instance.

aartaka commented:
How about the other way around: automagically look for exported symbols from other packages, whenever we can. This way, one can bind a key to `jump-to-heading`, meaning `nyxt/web-mode:jumpt-to-heading`. That would solve a lot of dependency problems and user configuration confusions.

Ambrevar commented:
I am not sure I understand: the point of packages is precisely to allow
for duplicate names without confusion.

If we automagically resolve `'jump-to-heading`, how can we disambiguate
between `nyxt/web-mode:jump-to-heading' and `nyxt-user/my-mode:jump-to-heading`?

We can however add a `find-nyxt-symbols` command that retrieves the
symbols from the packages of `nyxt`, `nyxt-user` and all their
subpackages.  Example use:

```lisp
(find-nyxt-symbols :jump-to-heading)
; => '(nyxt/web-mode:jump-to-heading nyxt-user/my-mode:jump-to-heading)
```

While the above is convenient at the REPL, one can see how it wouldn't
work programmatically because the program can't know which symbol to pick.

Ambrevar commented:
> solve a lot of dependency problems 

Do you have an example in mind?

Ambrevar commented:
Just realized something: `rememberable` and `togglers` go hand-in-hand, don't they?  Is there a case where a rememberable mode would not have a toggler and vice-versa?

If so, then we can have just one metaclass option, say `togglable-p`, and remove `rememberable`.  Then we need to update `auto-mode` accordingly.

Thoughts? @aartaka

aartaka commented:
> Another thought: further macrology could be removed from
> 
>     * `define-internal-page-command-global`
> 
>     * `define-internal-page-command`
> 
>     * `define-input-edit-command`
> 
>     * `define-bookmarklet-command`
> 
>     * `define-command-with-selection`
> 
>     * `define-panel`
> 
>     * `define-panel-global`
> 
> 
> by subclassing `command`. As usual, leveraging CLOS often proves a better idea than writing our own DSL. Thoughts? @aartaka

Yes!

aartaka commented:
> > solve a lot of dependency problems
> 
> Do you have an example in mind?

See the `help-mode` usage of `web-mode` commands noted by @aadcg.

Ambrevar commented:
Good point about `rememberable-p`, let's keep them both then.

Ambrevar commented:
> "invoke execute-command, look at the Mode column, and append nyxt/ to
> it -- you've got the real name of the command now!" It's clunky and
> opaque

Isn't it the opposite?  Seems to me that it's explicit and
straightforward.

We have describe-* commands to find all symbols across any package, so
we are never opaque.

> - Improve the documentation, init warnings, automatic configuration.

We can always do this, but what do you have in mind in particular about
the topic at hand?

> - Allow automatic resolution of _unambiguous_ symbols

I don't see how this can work.  What if the user / third-party extension
registers a symbol that's already in Nyxt?  It then becomes ambigous on
startup.

Example:

- Nyxt has a bunch of `(current-mode 'web-mode)` calls.
- The user defines "nyxt-user/my-package:web-mode".

Boom!  Nyxt won't work as expected.

Ambrevar commented:
I'm OK with having such resolution done in the init.lisp file, such as
in the example you gave with define-key.  However I don't think we
should use it in the core of Nyxt (I don't think it would be too useful either.)

This is aligned with my suggestion of how we could update
`current-mode`, .e.g `(current-mode :web)`.

What API do you suggest?  What about this:

```lisp
(current-mode :web)
; => <mode-class nyxt/web-mode:web-mode>
(find-command :jump-to-heading)`
; => `nyxt/web-mode:jump-to-heading'
(find-nyxt-symbol "element-focused")
; => `nyxt/web-mode:element-focused'
```

The 3 commands
- only find exported symbols in `nyxt`, `nyxt-user` or any subpackage;
- accept either a keyword or a (case-sensitive) string.
  Keyword is more convenient, but string is more universal (because of
  casing) plus it matches the API of `find-symbol`.

`find-command` is useful to avoid finding too ambiguities with
non-commands.  `find-nyxt-symbol' is the generic symbol bindings.

Thoughts?

Ambrevar commented:
@aartaka What do you think about the first question in the original post?
I was thinking we could remove `enable-mode` and instead make the MODES parameter optional in `enable-modes`, have it fall back to prompt the user if not set (like we do in `enable-mode`).

aartaka commented on source/command.lisp:
```diff
@@ -87,10 +87,19 @@ We need a `command' class for multiple reasons:
                  (if (eq :anonymous (visibility command)) nil (name command))
                  `(lambda ,arglist
-                    (declare (ignorable ,@(mapcar (alex:compose #'second #'first) keywords)))
+                    (declare (ignorable ,@(mapcar (alex:compose #'second #'first) keywords)
```
`cadar`?

Ambrevar replied:
Hmmm,,, It's shorter, but these ca*d*r functions are so hard to read...

aartaka commented:
> @aartaka What do you think about the first question in the original post?
> I was thinking we could remove `enable-mode` and instead make the MODES parameter optional in `enable-modes`, have it fall back to prompt the user if not set (like we do in `enable-mode`).

I don't see the downsides of it, but I don't see much use either :D But I guess one unambiguous function is better than two ambiguous ones!

Ambrevar commented:
> But why case-sensitive string? And why is there `NYXT/WEB-MODE:ELEMENT-FOCUSED` matching the `"element-focused"` string? Did you mean case-insensitive?

No, I meant case sensitive but forgot to uppercase the example :p  

The reason is to match `find-symbol`'s behviour plus allow for matching cased symbols like `|Foo|`.

> We can also accept a symbol and search it by `symbol-name`.

If we have keyword support, then symbol-name wouldn't be useful, plus it would intern an unbound symbol (which pollutes things like completion).

Ambrevar commented:
Your `resolve-symbol` is neat, thanks for sharing.  But maybe the type would be a bit too verbose in the general case?  We can always have `resolve-symbol` _and_ `find-command` for instance.

aartaka commented:
> Your `resolve-symbol` is neat, thanks for sharing. But maybe the type would be a bit too verbose in the general case? We can always have `resolve-symbol` _and_ `find-command` for instance.

We can make it optional and imply a `:command` or make some on-line helpers based on it, yes :)

Ambrevar commented:
Since we now depend on `swank-backend:arglist`, I suggest we pull in the https://github.com/Shinmera/trivial-arguments/ dependency to avoid explicit calls to swank.

We can also postpone this to later, it's a trivial matter.

Ambrevar commented:
This is mostly ready to merge.  Need to 
- test auto-mode.
- fix CCL.

I'll hopefully merge tomorrow, let me know if there is anything.

Ambrevar commented:
I'll keep the panel command rewrite for later.

Ambrevar commented:
Last minute questions.  Do you like the new API?

- `find-mode` is gone, only `find-submode` is used.  Rationale: `find-mode` was trivial (`(find 'mode-sym (modes buffer) :key 'class-name)`) and less general.  Could still be useful to find the exact mode instead of a submode, but for now it does not seem that we have a use case.
- The more convenient but less reliable `current-mode` is only meant to be used from the REPL, so it's no longer called in the source.
- `resolve-symbol` is handy and leverages `package-defined-symbols`, but the latter does not allow for listing symbols from _all_ packages.  Shouldn't we support the case we pass no package, which would act as "all packages"?
- The `mode-name` / `mode-symbol` / etc. craziness is gone.
- New `mode-class` function which returns the class associated to a symbol.  This is particularly useful to test if a symbol is that of a mode.

aartaka commented:
The API sounds sane! 

I'd still want to review the code (tommorrow, I suppose), as it's a huge change and we may have missed something :p

Ambrevar commented:
Sure, I'll be waiting :)

Ambrevar commented:
Artyom suggested to me that instead of calling `compile` to wrap the command so that it executes hooks, we could simply have a generic function...  Since they don't need to be specialize and have the hook system built-in, it would work!

That said, we need to find a way to generate anonymous generic functions, otherwise anonymous commands would be missing, which is critical in my opinion because they effectively parallel lambdas, which is at the core of Lisp / functional programming (closures, etc.).

Some hints to explore:

```lisp
(make-instance 'standard-generic-function :name 'foo :lambda-list '() :documentation "foo")

(c2mop:make-method-lambda #v31 (c2mop:generic-function-method-class #v31) (lambda () 17) nil )

(make-instance 'c2mop:standard-method :function (lambda () 17) )

(c2mop:make-method-lambda #v31 #v32 (lambda () 17) nil )
```

aartaka commented on source/mode/web.lisp:
```diff
@@ -747,3 +747,9 @@ ELEMENT-SCRIPT is a Parenscript script that is passed to `ps:ps'."
+
+(pushnew 'web-mode nyxt::%default-modes)
+
+(defmethod nyxt:default-modes :around ((buffer editor-buffer))
```
It's duplicated in buffer.lisp, isn't it?

Ambrevar replied:
I don't see it, maybe you misread?

This is not the prettiest relocation, but really we need to work on our buffer class architecture.  Next PR.

aartaka replied:
I have misread it, yes! Sorry ＜(。_。)＞

aartaka commented:
Phew, that's a huge one (=ｘェｘ=)

Ambrevar commented:
I've fixed auto-mode.  This raises a few questions:

- Commit 78ff70bc832cf366c74c0a2fb97ac1068b654eae indicates that auto-mode has been broken since forever for people who didn't set the request-resource-hook. Am I mistaken?
- In commit 319db27f1c48875ac7293d1d974e3e47d92786d0 I'm deprecating non-prefixed symbols.  They still work, but keywords/strings are preferred instead and won't trigger a warning.  Sound design?
- In commit 7a4c57b2d79c341071a811a644ba2034fca9a49b we must check that MODES is non-nil before passing it to `enable-modes`, because now that latter prompts the user for modes if the argument is NIL.  This is bad design... Looks like we need a lower level function after all.  Name suggestions?
- 6e9ff3e3361e447ec67e00873dabaa48720591b5 exhibits an odd race-condition which I've fixed, but I don't understand why this happens on this branch and not on `master`.

@aartaka ?

Ambrevar commented on source/command.lisp:
```diff
@@ -18,15 +18,26 @@ This is useful to build commands out of anonymous functions.")
+   (visibility :mode
+               :type (member :global :mode :anonymous)
+               :documentation "
+- `:global'  means it will be listed in `command-source' when the global option is on.
```
We do "pollute" Nyxt.  Misunderstanding?

Visibility serves the following:

- Have commands listed in command source even if the mode is not active.
- Have commands listed in the command source only if the mode is active.
- Don't list it.

What would you be suggesting otherwise?

aartaka replied:
We can do it via
- `nyxt`-interned/exported commands.
- Mode package exported commands.
- Unexported commands.

Ambrevar replied:
The original rationale behind controling visibilty other than via packages was so that third-party extensions can create global commands without polluting the `nyxt` package.

The solution of the `:visibility` slot is not as powerful as packages, sure, but it's a trivial implementation that's used in just 2-3 spots.  So I'd say yes, it's worth it, considering that the alternative means risking collisions with official commands.

Does that make sense?  Or did I misunderstand your proposal?

Ambrevar commented on source/command.lisp:
```diff
@@ -18,15 +18,26 @@ This is useful to build commands out of anonymous functions.")
+- `:mode' means the command is only listed in `command-source' when the corresponding mode is active.
+
+- `:anonymous' means the command is never listed in `command-source'.")
+   (deprecated-p nil
```
Yup, we've had had this TODO comment for a while:

```lisp
;; TODO: Should `define-deprecated-command' report the version
;; number of deprecation?  Maybe OK to just remove all deprecated
;; commands on major releases.
```

That said, this is older and not directly related to this pull request, so I suggest we address this after the merge.  Things are growing to far at this point :)

Ambrevar commented on source/command.lisp:
```diff
@@ -47,13 +58,96 @@ We need a `command' class for multiple reasons:
+    (eval-when (:compile-toplevel :load-toplevel :execute)
+      (let ((before-hook-sym (before-hook-name (name command)))
+            (after-hook-sym (after-hook-name (name command))))
+        (setf (symbol-value before-hook-sym) (before-hook command)
```
Why?  We never use `setq`.

Ambrevar commented on source/configuration.lisp:
```diff
@@ -346,3 +346,34 @@ See `on'."
+                    (*package* (symbol-package (swank-backend:function-name fun)))
+                    (file-content (alexandria:read-file-into-string (first (alexandria:assoc-value definition :file))))
+                    (start-position (first (alexandria:assoc-value definition :position))))
+      (restart-case
```
Yes, but for other errors the stack would be rewound, with no possibility of restart.

In general, I propose we prefer `handler-bind` to `handler-case` because it allows for much better debugger use.

Ambrevar commented on source/mode.lisp:
```diff
@@ -3,271 +3,432 @@
+        (make-instance
+         'command
+         :name name
+         :visibility :global
```
Some day, some day...

Ambrevar commented on source/mode.lisp:
```diff
@@ -3,271 +3,432 @@
+    (error "Cannot initialize `mode', you must subclass it.")))
+
+(export-always 'enable)
+(defgeneric enable (mode &key &allow-other-keys)
```
No, because now modes are never destroyed.  We can disable them, but not destroy them.  If destruction is ever needed, then we can add a `destroy` function, but that's a separate concern.  

See https://github.com/atlas-engineer/nyxt/issues/1005#issuecomment-1015202688 for a rationale.

Ambrevar commented on source/mode.lisp:
```diff
@@ -3,271 +3,432 @@
+  "Shorthand to define a mode.  It has the same syntax as `define-class' but
+optionally accepts a docstring after the superclass declaration.
+The `mode' superclass is automatically added if not present."
+  (let* ((docstring (when (stringp (first body))
```
I was wondering the same, but `parse-body` is for function bodies, and here we have a "class body", which is different because of the slots and the class options.

Ambrevar commented on source/mode.lisp:
```diff
@@ -3,271 +3,432 @@
+                                (sera:filter
+                                 (apply #'alex:disjoin
+                                        (mapcar (lambda (pkg)
+                                                  (alex:rcurry #'subpackage-p (find-package pkg)))
```
Quite normal, I've added it :)

Ambrevar commented on source/mode.lisp:
```diff
@@ -3,271 +3,432 @@
+  ((prompter:name "Inactive modes")
+   (buffers :initarg :buffers :accessor buffers :initform nil)
+   (prompter:multi-selection-p t)
+   (prompter:constructor (lambda (source)
```
Constructor effectively `constructs` a source.  Prompt buffer  sources cannot be enabled or disabled.  They are just used, then destroyed.

aartaka replied:
Yes, does make sense!

Ambrevar commented on source/mode/blocker.lisp:
```diff
@@ -83,23 +83,21 @@ Example:
+   (load-hostlists-p t)))
+
+(defmethod enable ((mode blocker-mode) &key)
+  (when (web-buffer-p (buffer mode))
```
I think it's been obsoleted by my recent rewrite of the mode, it's just a forgotten left-over.

I'll double-check.

Ambrevar commented on source/mode/editor.lisp:
```diff
@@ -23,21 +23,18 @@ get/set-content (which is necessary for operation)."
   (let* ((content (spinneret:with-html-string
-                   (:head (:style (style (buffer editor))))
-                   (:body (:p "Please configure an editor mode to use an editor buffer."))))
+                    (:head (:style (style (buffer editor))))
```
We can fix it later.  I'll add a TODO for now.

Ambrevar commented on source/mode/emacs.lisp:
```diff
@@ -21,13 +21,13 @@ Example:
+this mode.")))
+
+(defmethod enable ((mode emacs-mode) &key)
+  (with-accessors ((buffer buffer)) mode
```
I'm unsure about this in terms of good practice.

If we have an accessor, we should use it by default _unless_ there is an explicit need to bypass the accessor.  THoughts?

aartaka replied:
Given that we entered the unexplored land of methods with lots of qualifiers and combinations now, it does make sense sticking to accessors -- who knows how crazy of a difference can there be between the slot and the accessor result ¯\\(°_o)/¯

Ambrevar commented on source/mode/expedition.lisp:
```diff
@@ -54,5 +54,5 @@
          (urls (mapcar #'quri:uri urls))
          (buffer (make-buffer :title "" :url (first urls))))
-    (expedition-mode :urls urls :buffer buffer)
+    (enable (make-instance 'expedition-mode :urls urls :buffer buffer))
```
It does, this is a forgotten draft, I'll fix it.

Ambrevar replied:
Actually, both have their benefits.  The command does not list the arguments, while the `make-instance` does.

aartaka replied:
I see. And also, some modes don't have toggler commands anymore, so we're saver with `enable`.

Ambrevar commented on source/mode/no-image.lisp:
```diff
@@ -8,9 +8,10 @@
-   (constructor
-    (lambda (mode)
-      (ffi-buffer-auto-load-image (buffer mode) nil)))))
+  ())
```
I wonder... `defclass` forbids this, although I don't know why.  Shan't we be consistent with defclass?

aartaka replied:
I suppose that it's just a historic accident that `defclass` forbids it. If we can do it shorter, why not do it?

aartaka commented:
> * Commit [78ff70b](https://github.com/atlas-engineer/nyxt/commit/78ff70bc832cf366c74c0a2fb97ac1068b654eae) indicates that auto-mode has been broken since forever for people who didn't set the request-resource-hook. Am I mistaken?

Oh wow :D I haven't checked it with an empty `request-resource-hook`, so I cannot tell. But I remember that it worked without one when I was writing it (it was some two years ago, so don't trust my word here).

> * In commit [319db27](https://github.com/atlas-engineer/nyxt/commit/319db27f1c48875ac7293d1d974e3e47d92786d0) I'm deprecating non-prefixed symbols.  They still work, but keywords/strings are preferred instead and won't trigger a warning.  Sound design?

Hmmmmmm. It does make sense and sounds good. But oh man, how much rules I'll have to rewrite!

> * In commit [7a4c57b](https://github.com/atlas-engineer/nyxt/commit/7a4c57b2d79c341071a811a644ba2034fca9a49b) we must check that MODES is non-nil before passing it to `enable-modes`, because now that latter prompts the user for modes if the argument is NIL.  This is bad design... Looks like we need a lower level function after all.  Name suggestions?

Maybe use the supplied-p argument to check whether the nil comes from the user or the interactive use?

> * [6e9ff3e](https://github.com/atlas-engineer/nyxt/commit/6e9ff3e3361e447ec67e00873dabaa48720591b5) exhibits an odd race-condition which I've fixed, but I don't understand why this happens on this branch and not on `master`.

0_o

jmercouris commented on source/mode/reading-line.lisp:
```diff
@@ -41,18 +41,12 @@ mode."
 (define-command jump-to-reading-line-cursor (&key (buffer (current-buffer)))
   "Move the view port to show the reading line cursor."
   (pflet ((jump-to-cursor ()
-    (ps:chain (ps:chain document (query-selector "#reading-line-cursor")) (scroll-into-view-if-needed))))
```
Yes, we should can we put it into the locals?

Ambrevar commented:
OK to merge then?

aartaka commented:
Yes!

aartaka commented:
Oh wait, CCL tests fail :P

Ambrevar commented:
I'll fix it.

There is one relatively important regression: "go to source" does not work anymore for commands.  Try typing `(execute-command )` in a REPL, press "M-." on the symbol and it'll fail to go to the source.

I don't understand why this is happening, any idea?

I can call

```lisp
(eval-when (:compile-toplevel) (sb-c:%compiler-defun ',name t nil nil))
```

and it works (but that fails to compile because of double-defs).  Need to get to the bottom of this.

Ambrevar commented:
I've narrowed it down to the following.  Produce the following snippet in a file and compile it:

```lisp

(defun kaboom ()
  1)
(let ((comp (compile 'kaboom '(lambda () 2))))
  (setf (fdefinition 'kaboom) (uiop:ensure-function comp)))
```

Now you can run `kaboom` but the REPL / SBCL fails to find it's location.

Note that if you remove the quote before the lambda, then it works!

Quoted lambdas are supported, although it's not well documented.  From CLHS:

```
Examples:
[...]
 (setf (symbol-function 'foo)
       (compile nil '(lambda () "replaced"))) =>  #<Compiled-Function>
 (foo) =>  "replaced"
```

But Nyxt needs the backquote so that it can generate the argument list.

Any idea?

Ambrevar commented:
I've made progress on the anonymous generic function front:

```lisp
(c2mop:ensure-generic-function 'foogf)
(fmakunbound 'foogf)
(c2mop:ensure-method ** '(lambda () 17))
(c2mop:ensure-method ***  '(lambda () (print 'ok)) :qualifiers '(:after))

(funcall #v0)

OK
; => 17
```

The `fmakunbound` trick is a bit clunky, but it works!

Alternative, the "pure" version works too on SBCL:

```lisp
(make-instance 'standard-generic-function :documentation "docstring" :name 'foogf)
```

Just got to make sure this is portable.  Could we leverage CLOS here?

Ambrevar commented:
Next question: can `command` subclass `standard-generic-function`?  That'd be awesome!

Ambrevar commented:
And here is a draft:

```lisp
(defclass command (standard-generic-function)
  ((deprecated-p
    :initform nil
    :accessor deprecated-p
    :initarg :deprecated-p))
  (:metaclass sb-mop:funcallable-standard-class))

(defmethod initialize-instance :after ((cmd command) &key)
  ;; TODO: Can we compute the lambda list dynamically?  With
  ;; `closer-mop:generic-function-lambda-list'?
  
  ;; Alternatively, don't add a method but instead specialize a closer-mop
  ;; generic function that's related to method computation.
  (c2mop:ensure-method
   cmd
   `(lambda ()
      (when (deprecated-p ,cmd)
        (log:info "deprecated")))
   ;; TODO: :before or :around?
   :qualifiers '(:before)))

(make-instance 'command
               :name 'foogf
               :documentation "docstring" 
               :deprecated-p t)

(c2mop:ensure-method * '(lambda () (log:info "OK")) )
```

aartaka commented:
I knew we can do it! Yay :D

aartaka commented:
Does it allow go-to-source?

Ambrevar commented:
Just like regular defgeneric / defmethod I presume!

I really like this approach.  If it works, it's going to be both simpler / shorter and much more powerful than our current `command` implementation thanks to the extensive framework around generic functions in MOP / CLOS!

Thanks for pushing me to look towards this direction, @aartaka 

Today I understood something of which I had failed to acknowledge the depth over the past few years: generic functions in Lisp are not called `generic` (just) because they apply to different types of arguments (with multiple dispatch) but most importantly because they are _generic and extensible_ as a first class concept!

Ambrevar commented:
Better way to deal with the `deprecated` option?

```lisp

(defmethod closer-mop:compute-effective-method :around ((cmd command) combin applicable)
  (when (deprecated-p cmd)
    (log:warn "Deprecated"))
  (call-next-method))
```

The benefit is that we don't need to fiddle with `initialize-instance :after` nor disturb the actual `:before` or `:around` methods, which the user may overwrite.

Also this approach makes many other things much simpler.  For instance, `define-command` is now as simple as

```lisp

(defmacro define-command (...)
  `(defgeneric my-cmd (args)
     (:generic-function-class command)))
```

And `define-deprecated-command` (if needed at all, considering how trivial it is to deprecate it now):

```lisp

(defmacro define-deprecated-command (...)
  `(prog1 (defgeneric my-cmd (args)
            (:generic-function-class command))
     (setf (deprecated-p #'my-cmd) t)))
```

Ambrevar commented:
Anyways, how shall we proceed?  We can merge this now and I'll open a new pull request to turn commands into generic functions.

Agreed?

aartaka commented:
Yes, that can wait for the next PR, I guess :)

Ambrevar commented:
Merged!  Brace yourself for a tsunami of bugs :D

aartaka commented:
＼（＠￣∇￣＠）／

Ambrevar commented:
I'm OK with having such resolution done in the init.lisp file, such as
in the example you gave with define-key.  However I don't think we
should use it in the core of Nyxt (I don't think it would be too useful either.)

This is aligned with my suggestion of how we could update
`current-mode`, .e.g `(current-mode :web)`.

What API do you suggest?  What about this:

```
(current-mode :web)
; => <mode-class nyxt/web-mode:web-mode>
(find-command :jump-to-heading)`
; => `nyxt/web-mode:jump-to-heading'
(find-nyxt-symbol "element-focused")
; => `nyxt/web-mode:element-focused'
```

The 3 commands
- only find exported symbols in `nyxt`, `nyxt-user` or any subpackage;
- accept either a keyword or a (case-sensitive) string.
  Keyword is more convenient, but string is more universal (because of
  casing) plus it matches the API of `find-symbol`.

`find-command` is useful to avoid finding too ambiguities with
non-commands.  `find-nyxt-symbol' is the generic symbol bindings.

Thoughs?

aartaka commented:
I don't understand who this commentary directed to and in which context 0_o

jmercouris commented:
Probably the mail client is a little confused. Pierre where did you intend to send this?

Ambrevar commented:
Nope, what the hell is this!?!

Probably a GitHub bug, or like John said an email bug.
